### ISSUE TITLE
Crash regarding xcb on Linux, Java 11+. Also Sound is buggy and doesn't work sometimes.

### ISSUE THREAD

**Issue by Lachrymogenic** [MinicraftPlus/minicraft-plus-revived#681]
**Describe the bug**
The game runs with no audio on OpenJDK Java 8, so I upgraded to 11, 17 and 21 and the game kept crashing with the exact same error message, that message being: 
```
[Game Handler] WARN: Game tried to exit display, but no menu is open.
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
java: ../../src/xcb_io.c:269: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
```
I am using OpenJDK.
I downloaded an official Java 8 binary from java.com and the game now works completely fine however, in ``pavucontrol`` the volume slider for the game is flickering. Heres a video (Java version here is ``Java(TM) SE Runtime Environment (build 1.8.0_411-b09)
``):

[strange.webm](https://github.com/MinicraftPlus/minicraft-plus-revived/assets/160751404/39fe5231-97b5-44fc-acfc-4ab6a98a997b)

I genuinely find this to be so weird, even the recording software is not picking up the sound correctly.

**To Reproduce**
I have no idea if it is reproducable, it could be an issue with my machine but then again, what am I supposed to do about it? Other java applications like Minecraft work flawlessly on OpenJDK Java 8 and other versions like OpenJDK Java 17. So I'm pretty sure something is wrong with the game itself or it was not meant for anything except Java 8. If that's the case, I have no idea why sound isn't working on Java 8. Here's a video of sound not working on ``OpenJDK Runtime Environment (build 1.8.0_402-8u402-ga-2ubuntu1~20.04-b06)``

https://github.com/MinicraftPlus/minicraft-plus-revived/assets/160751404/c7488105-d3f9-492e-b01c-bdf20d16e14c

Albeit this is a bad example because I didn't play any example audio from a browser or something, so the video just has no audio, but you can see from the pavucontrol that the java stream is not producing any audio.
**Expected behavior**
No crash. (and for sound to work)

**Desktop (please complete the following information):**

- OS: Ubuntu 20.04.6 LTS x86_64 
- Game version: 2.2.0

**Additional Comments**:
I personally don't know why this happens, even tried building on my own machine and same thing occurs. Sound works but stutters in Java 11+ and crashes with that error.

**Comment by Lachrymogenic** [comment]
After closer inspection, you can see that the audio that does work is ALSA, but when it doesn't, it's just "Audio Stream"
For some reason that specific OpenJDK version isn't connecting to ALSA, or perhaps the issue is ALSA?

**Comment by Lachrymogenic** [comment]
After playing more on that specific Java version, Java(TM) SE Runtime Environment (build 1.8.0_411-b09), the audio became significantly delayed, I did have a youtube video playing in the background is that makes a difference.

**Comment by Litorom** [comment]
@BenCheung0422 @Makkkkus

**Comment by Lachrymogenic** [comment]
Unfortunately, Java(TM) SE Runtime Environment (build 1.8.0_411-b09) still causes the xcb crash (same error message btw), but it's very inconsistent, sometimes I can play for more than 10 to 20 minutes and not encounter an issue and sometimes like just now it crashed within 5 minutes. I have no idea if xcb is an issue, but I can't seem to find a solution.

**Comment by BenCheung0422** [comment]
Furthermore, about the audio stream names:

Before the change to the audio system mentioned, there are 10 streams all named "Java: Audio Clip", exactly as the number of audio files in game, as intended.

After the change, 1.8 does indeed gives a single "Java: Audio Stream" audio stream and other versions give a "ALSA plug-in [java]: ALSA Playback" stream. I apologize that I missed the information that you have also tested with Oracle JRE 8, and it can indeed play audio with delays (seemingly with smaller delays), but I still cannot notice of the "flicks" of stream. It seems like OpenJDK JRE 8 is not linking with ALSA.

In conclusion, I am not sure whether this issue is reported or occurred only during some specific circumstances. And maybe tests in other Linux may be conducted. I will try to debug with using Ubuntu if possible.

**Comment by Lachrymogenic** [comment]
Seems to work on my Steam Deck just fine, using SteamOS Holo 3.5.19 with Openlogic OpenJDK Java 8 JRE 8u352-b08-linux-64. Maybe that's because Steam Deck uses pipewire and not pulseaudio?

**Comment by BenCheung0422** [comment]
If that is the case that there are some issues with PulseAudio, we might have to check with it.

**Comment by BenCheung0422** [comment]
Can you try the solution mentioned [here (stackoverflow)](https://stackoverflow.com/a/46121673)?

**Comment by Lachrymogenic** [comment]
I have tried this and audio is now playing on my stock OpenJDK 8 but pavucontrol still flickers for some reason.
The audio plays but it's a little jittery, maybe I can show you a video.

https://github.com/MinicraftPlus/minicraft-plus-revived/assets/160751404/00e2fa7e-4fda-41d2-bcac-43cd73d44835

The jitteryness is only noticeable when watching playback of a video, otherwise it's only partially noticeable 
My stock OpenJDK 8 version is ``OpenJDK Runtime Environment (build 1.8.0_402-8u402-ga-2ubuntu1~20.04-b06)
``
Reverting sound.properties back to normal gives no sound, so I'm guessing it's an issue with PulseAudio.

**Comment by BenCheung0422** [comment]
It looks like there is some compatibility issues between PulseAudio and Java, and indeed OpenJDK 8 PulseAudio API/IcedTea is broken. The unstable connection between PulseAudio and Java may be caused by certain configurations, but more tests shall still be conducted.

**Comment by BenCheung0422** [comment]
My test with OpenJDK 11

https://github.com/MinicraftPlus/minicraft-plus-revived/assets/74168521/89b55c93-277f-47e2-9450-4e0dc1e505d4

**Comment by BenCheung0422** [comment]
My suggestion is to use PipeWire over PulseAudio. PipeWire is a newer package than PulseAudio, and the compatibility should be better.

**Comment by BenCheung0422** [comment]
Another way is to reset the sound configurations to the default settings:
```
rm -f ~/.config/pulse/*
sudo killall pulseaudio
# daemon(s) will restart imediately
```
More Info: https://askubuntu.com/a/1259763

And you may try to spot out the problematic configs by reverting it to be the ones before resetting.

**Comment by Lachrymogenic** [comment]
> My suggestion is to use PipeWire over PulseAudio. PipeWire is a newer package than PulseAudio, and the compatibility should be better.

I'm not doing that, I've tried setting it up and it's just a pain, especially on Ubuntu it seems.
There's no reason why other Java applications can be perfect with audio with PulseAudio, but this one cannot?
For example, Minecraft works perfectly fine with PulseAudio on my system, so why should Minicraft have issues?

I do get that they're not the same or whatever, but I'm just wondering if a new sound system could solve this issue, as it shouldn't be up to the user to make changes to their system to have a game work as intended, the program should work on all systems and be compatible with not only pipewire but pulse too. Essentially, literally everything else works fine with Pulse so why should I have to switch just to play Minicraft, that just doesn't make any sense to me.

I'll just deal with the flickering sound for now, because it is better than nothing.

**Comment by BenCheung0422** [comment]
The sound system we are currently using heavily relies on native Java sound libraries while Mincraft's LWJGL is using OpenAL for audio. I doubt that the Java native libraries have some problems tackling with audio on some Linux distributions.

**Comment by BenCheung0422** [comment]
However, more than that, my 22.04 Ubuntu VM test (no custom settings) with PulseAudio does not come with flickering issue. I also guess that Java audio may not be compatible with certain configurations.

**Comment by Lachrymogenic** [comment]
> However, more than that, my 22.04 Ubuntu VM test (no custom settings) with PulseAudio does not come with flickering issue. I also guess that Java audio may not be compatible with certain configurations.

What OpenJDK versions do you have on the VM and what is in the sound.properties file?
Is it just stock or normal?

**Comment by BenCheung0422** [comment]
The following is the setups and results with my tests:
- `OpenJDK Runtime Environment (build 1.8.0_402-8u402-ga-2ubuntu1~22.04-b06)` with modified `sound.properties` (mentioned above)
- `OpenJDK Runtime Environment (build 11.0.22+7-post-Ubuntu-0ubuntu222.04.1)` with default `sound.properties` (empty)
- `OpenJDK Runtime Environment (build 17.0.10+7-Ubuntu-122.04.1)` with default `sound.properties` (empty)
- `OpenJDK Runtime Environment (build 21.0.2+13-Ubuntu-122.04.1)` with default `sound.properties` (empty)

They however sometimes come with significant audio latencies.

**Comment by zandgall** [comment]
Discovered to be fixed by using OpenAL or some wrapper equivalent. Putting the switch to use OpenAL on my list of things to do.
